kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/MemoryRecordsUtils.java

```diff
@@ -65,12 +73,42 @@ private static BatchAwareMemoryRecordsBuilder appendAll(BatchAwareMemoryRecordsB
 
     /**
      * Factory method for a Collector that concatenates MemoryRecords.
+     * @param onClose
+     * @param resultBuffer
      */
-    public static @NonNull Collector<MemoryRecords, BatchAwareMemoryRecordsBuilder, MemoryRecords> concatCollector(@NonNull ByteBufferOutputStream resultBuffer) {
-        return Collector.of(
-                () -> new BatchAwareMemoryRecordsBuilder(resultBuffer),
-                MemoryRecordsUtils::appendAll,
-                MemoryRecordsUtils::combineBuilders,
-                BatchAwareMemoryRecordsBuilder::build);
+    public static @NonNull Collector<MemoryRecords, BatchAwareMemoryRecordsBuilder, MemoryRecords> concatCollector(
+                                                                                                                   Consumer<Runnable> onClose,
+                                                                                                                   @NonNull ByteBufferOutputStream resultBuffer) {
+        Objects.requireNonNull(resultBuffer);
+        return new Collector<>() {
+            @Override
+            public Supplier<BatchAwareMemoryRecordsBuilder> supplier() {
+                return () -> {
+                    BatchAwareMemoryRecordsBuilder batchAwareMemoryRecordsBuilder = new BatchAwareMemoryRecordsBuilder(resultBuffer);
+                    onClose.accept(batchAwareMemoryRecordsBuilder::close);
```

Comment:
it wasn't immediately obvious to me why you've separated the responsibility to close, rather than give it to the finisher.   Maybe some javadoc would help.

Reply:
I'll add some, but ftr:

1. the finisher is not guaranteed to be called (e.g. in the event of an exception being thrown).
2. In a parallel stream the finisher is only called for the final intermediate result, not for any intermediates that arose from partitioning the underlying Spliterator. On the happy both those will be closed by the combiner, but wouldn't be closed if exceptions are thrown.

We're not expecting streams backed by our stream factory methods to be parallel, so 2 probably doesn't apply, but 1 definitely does. I suppose it's also possible for the `Collector`s in this PR to be used with `Streams<Record>` from other, parallel, sources.

Reply:
It was sufficiently difficult to document that I made this method private and added a public method to do the concat.